src/programs/azimuthal_average/azimuthal_average.cpp

```diff
@@ -0,0 +1,1534 @@
+	float acceleration_voltage = my_current_job.arguments[2].ReturnFloatArgument();
+	float spherical_aberration = my_current_job.arguments[3].ReturnFloatArgument();
+	float amplitude_contrast = my_current_job.arguments[4].ReturnFloatArgument();
+	float defocus_1 = my_current_job.arguments[5].ReturnFloatArgument();
+	float defocus_2 = my_current_job.arguments[6].ReturnFloatArgument();
+	float astigmatism_angle = my_current_job.arguments[7].ReturnFloatArgument();
+	float additional_phase_shift = my_current_job.arguments[8].ReturnFloatArgument();
+	bool input_ctf_values_from_text_file = my_current_job.arguments[9].ReturnBoolArgument();
+	std::string text_filename = my_current_job.arguments[10].ReturnStringArgument();
+	bool phase_flip_only = my_current_job.arguments[11].ReturnBoolArgument();
+	int max_iterations = my_current_job.arguments[12].ReturnIntegerArgument();
+	float minimum_shift_in_angstroms = my_current_job.arguments[13].ReturnFloatArgument();
+	float maximum_shift_in_angstroms = my_current_job.arguments[14].ReturnFloatArgument();
+	float termination_threshold_in_angstroms = my_current_job.arguments[15].ReturnFloatArgument();
+	float psi_min = my_current_job.arguments[16].ReturnFloatArgument();
+	float psi_max = my_current_job.arguments[17].ReturnFloatArgument();
+	float psi_step = my_current_job.arguments[18].ReturnFloatArgument();
+	float termination_threshold_in_degrees = my_current_job.arguments[19].ReturnFloatArgument();
+	float bfactor_in_angstroms = my_current_job.arguments[20].ReturnFloatArgument();
+	bool should_mask_central_cross = my_current_job.arguments[21].ReturnBoolArgument();
+	int horizontal_mask_size = my_current_job.arguments[22].ReturnIntegerArgument();
+	int vertical_mask_size = my_current_job.arguments[23].ReturnIntegerArgument();
+	float exposure_per_frame = my_current_job.arguments[24].ReturnFloatArgument();
+	int number_of_frames_for_running_average = my_current_job.arguments[25].ReturnIntegerArgument();
+	int max_threads = my_current_job.arguments[26].ReturnIntegerArgument();
+
+	// The Files
+	ImageFile my_input_file(input_filename, false);
+	NumericTextFile *input_text;
+	long number_of_input_images = my_input_file.ReturnNumberOfSlices();
+
+	if (input_ctf_values_from_text_file == true)
+	{
+		input_text = new NumericTextFile(text_filename, OPEN_TO_READ);
+		if (input_text->number_of_lines != number_of_input_images)
+		{
+			SendError("Error: Number of lines in defocus text file != number of images!");
+			DEBUG_ABORT;
+		}
+
+		if (input_text->records_per_line != 3 && input_text->records_per_line != 4)
+		{
+			SendError("Error: Expect 3 or 4 records per line in defocus text file");
+			DEBUG_ABORT;
+		}
+	}
+
+	// CTF object
+	// temporary array to hold CTF parameters read from file
+	float temp_float[5];
+	// allocate 2D array for CTF
+	int ncols_ctf = 11;
+	// dynamically create array of pointers of size number_of_images (rows)
+	float** ctf_parameters = new float*[number_of_input_images];
```

Comment:
I think an array of structures here would be less prone to mistakes, i.e. ctf_parameteres.defocus1 etc.

Reply:
I thought about a more elegant way of storing the ctf parameters before eventually deciding on using a dynamically allocated 2d array as a way to get a working code. Using a C++ approach, I was thinking of creating a ctf_parameter struct and defining a vector of type ctf_parameter and then storing each image's ctf as a struct and adding it to a vector called ctf_parameter_stack. I am not too familiar with using c for dynamic data structures, but I will look more into it.